_posts/2022-06-20-qute-templates-from-db.adoc

```diff
@@ -0,0 +1,282 @@
+
+The next sections will make use of this JPA entity:
+
+[source,java]
+----
+package org.acme;
+
+import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
+
+import javax.persistence.Entity;
+import javax.persistence.Id;
+
+@Entity
+public class DbTemplate extends PanacheEntityBase {
+
+    @Id
+    public String name; <1>
+
+    public String content;
+}
+----
+<1> The template name will be the DB primary key.
+
+== The interesting part: connecting Qute to the database
+
+Now that templates can be persisted, we need a way to use them from Qute.
+Fortunately, Qute comes with a very interesting interface called `TemplateLocator` that can be used to load templates from any location, including from a DB.
+
+Here's how it can be used with the `DbTemplate` entity we defined earlier:
+
+[source,java]
+----
+package org.acme;
+
+import io.quarkus.logging.Log;
+import io.quarkus.qute.TemplateLocator;
+import io.quarkus.qute.Variant;
+
+import javax.enterprise.context.ApplicationScoped;
+import java.io.Reader;
+import java.io.StringReader;
+import java.util.Optional;
+
+@ApplicationScoped
+public class DbTemplateLocator implements TemplateLocator {
+
+    @Override
+    public Optional<TemplateLocation> locate(String name) {
+        DbTemplate template = DbTemplate.findById(name);
+        if (template == null) {
+            Log.tracef("Template with [name=%s] not found in the database", name);
+            return Optional.empty();
+        } else {
+            Log.tracef("Template with [name=%s] found in the database", name);
+            return Optional.of(buildTemplateLocation(template.getContent()));
+        }
+    }
+
+    @Override
+    public int getPriority() { <1>
+        return DEFAULT_PRIORITY - 1;
+    }
+
+    private TemplateLocation buildTemplateLocation(String templateContent) {
+        return new TemplateLocation() {
+
+            @Override
+            public Reader read() {
+                return new StringReader(templateContent);
+            }
+
+            @Override
+            public Optional<Variant> getVariant() {
+                return Optional.empty();
+            }
+        };
+    }
+}
+----
+<1> If your Quarkus app contains templates loaded from both the file system and a database, you will need to override the template locator default priority. Otherwise, Quarkus will try to load templates from the file system using `DbTemplateLocator` which could lead to exceptions or unpredictable behaviors.
+
+Before Quarkus 2.10, integrating `DbTemplateLocator` with the Qute engine instance provided by Quarkus could only be done with a CDI observer:
+
+[source,java]
+----
+package org.acme;
+
+import io.quarkus.qute.EngineBuilder;
+
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+
+public class QuteEngineConfiguration {
+
+    @Inject
+    DbTemplateLocator dbTemplateLocator;
+
+    void configureEngine(@Observes EngineBuilder builder) {
```

Comment:
In fact, you could try to add this observer method directly to the `DbTemplateLocator`, i.e.:
```java
void configureEngine(@Observes EngineBuilder builder) {
   builder.addLocator(this);
}
```

Reply:
I updated the post with that suggestion, thanks!